[RFC] SOTA-System: Fills architecture documentation using arc42 template

FlorentRevest opened the pull request:
Hi,

This is my first draft of a SOTA System documentation using the arc42 template. There might be a couple of problems here and there (for example with some diagrams written using blockdiag).

I believe there still aren't enough use cases for the Runtime View paragraph. If you have ideas of use cases feel free to share! 

I tried to stick to the template as much as possible but I may have modified a couple of paragraphs I judged inappropriate for our needs.

The current state of the page also misses some paragraphs on more pragmatic issues such as how to install Hawkbit or how to install a .swu from the CLI. Would you have an idea of where I could put that? I thought of adding other appendixes.

Any feedback is welcome! Thanks

FlorentRevest committed:
SOTA-System: Fills architecture documentation using arc42 template

The SOTA-Client page has been renamed to SOTA-System to encompass both
SWUpdate and Hawkbit.

This documentation tries to follow the arc42 template as defined by:
http://arc42.org/ Some freedom have been taken here and there.

Signed-off-by: Florent Revest <revestflo@gmail.com>

GordanM commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+
+The SOTA System is responsible for distributing and applying updates using
+over-the-air mechanisms so that PELUX-running machines can get new
+functionalities and security fixes.
```
s/functionalities/functionality

FlorentRevest replied:
Eheh, I wanted to check if this was idiomatic and found:

"Functionalities is not wrong, it's just excessive. It sounds like business jargon, and business jargon is often used by people who are better at making money than writing coherent prose."

I'm doomed!

GordanM commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+Basic requirements
+""""""""""""""""""
+
+A SOTA system should allow Over-The-Air updates of PELUX installations. Given
```
`Over-The-Air` was written in small caps in the `Introduction and goals` chapter. It would be nice to have it consistent :).

GordanM commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+^^^^^^^^^^^^^^^^
+
+Update artifacts generated by the build server, stored by the SOTA Server and
+fetched and applied by the SOTA Client are bundled in a format named ".swu". It
```
Should the format be in `code` brackets? @tobsan do you know the convention we are using?

jonte replied:
You can use touble backticks, like ``this``

tobsan replied:
I think we use ``backticks`` for most such text, yes.

GordanM commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+fetched and applied by the SOTA Client are bundled in a format named ".swu". It
+is essentially a CPIO archive containing various files and scripts and most
+importantly a top-level "sw-description" describing the content of the bundle.
+This sw-description must be written by hand by PELUX developers in a XML or
```
What is a "sw-description"? Is this a term commonly used in this context (SWupdate)?

FlorentRevest replied:
I clarified that it was a file. :)

GordanM commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+example recipes to integrate an image to a couple of platforms.
+
+meta-swupdate has been integrated to the default PELUX manifests, to avoid code
+duplication. However, it has been decided that the meta-swupdate-boards
```
Nice that you have mentioned that. What is the reason behind that, how is the code superfluous, maybe a sentence about it would clarify more :).

GordanM reviewed (commented):
Very nice job! I've just added a few nitpick comments :)

jonte commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
++----------------+-------------------------------------------------------------+
+| RA-2           | Automatic revert when a non-working update is applied       |
++----------------+-------------------------------------------------------------+
+| RA-3           | Updates integrity and signature verification                |
```
RA-3 is a bit unclear to me - can it be rewritten? I think you mean that "Integrity and signatures of updates are verified", or something of that sort?

jonte commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+^^^^^^^^^^^^^
+
++----------+--------------+----------------------------------------------------+
+| Priority | Quality-Goal | Scenario                                           |
```
I think we should also add efficiency goals, like how fast an update can be applied. I'm not saying it has high prio - but having it in here shows which prio we believe it has.

FlorentRevest replied:
This is definitely a good idea but I'm afraid I wouldn't know what kind of time target could be expected from such a system.

jonte commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+^^^^^^^^^^^^
+
++-----------------------+--------------------------+---------------------------+
+| Role                  | Description              | Goal, Intention           |
```
What about the user of the SOTA software in the car?

FlorentRevest replied:
I am not sure how/if the user is supposed to interact with the SOTA system. This sounds like a UX problem to me.
Some GUI implementations could decide to apply the updates by themselves or let the user decide if he wants to apply updates later. In each case, I would argue that the user only interacts with the GUI and the GUI interacts with the SOTA system itself as described in the first row of this table.

But maybe I am misunderstanding! :)

jonte commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
++------------------+-----------------------------------------------------------+
+| GUI              | Uses the SOTA Client to trigger an update                 |
++------------------+-----------------------------------------------------------+
+| Bootloader       | Used by the SOTA Client to atomically update a system     |
```
Isn't the bootloader mainly used to select the system to boot (in a partition A/B setup, for instance)? Does it also perform the actual update?

jonte commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
++------------------+-----------------------------------------------------------+
+| Partitions       | Used by the SOTA Client to update a rootfs                |
++------------------+-----------------------------------------------------------+
+| ECUs/FPGAs       | Used by the SOTA Client to update satellite processing    |
```
May want to call these devices "offboard" rather than satellite, just for clarity ;)

jonte reviewed (commented):
Use of arc42 template looks good to me!

thenor commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,614 @@
+----------------------
+
+The SOTA System is responsible for distributing and applying updates using
+over-the-air mechanisms so that PELUX-running machines can get new
```
I see that you have used this as a sort of "identifier" later on, but I still feel that I would like you to rewrite this as "machines running PELUX".
Maybe others disagree.

GordanM reviewed (commented):
One more comment about the second commit, the shortlog is too long and Dictionnary is written `Dictionary`. :)

FlorentRevest commented:
By the way, what is the recommended way of adding keywords to the spelling_wordlist.txt file: in a separate commit or integrated with the other commit ?

I had my previous PR merged with one commit for the actual documentation content and another commit for the dictionary. I thought it was a good decision because each commit was kept atomic.

But at the same time, it breaks tests bisectability which is a bad habit. I also noticed from the logs of aspell.en.pws that most other modifications were gathered in a single commit.

Should I squash my two commits into one?

FlorentRevest committed:
Use sphinxcontrib-plantuml for UML diagrams drawings

Blockdiag isn't enough to draw complex UML diagrams. PlantUML does the
job much better and even has nice companion tools such as
plantumlqeditor.

More details at: http://plantuml.com/

Signed-off-by: Florent Revest <revestflo@gmail.com>

FlorentRevest commented:
I upgraded my PR with:
- Every UML diagrams rewritten (and largely enhanced) with PlantUML. This is definitely the way to go when writing UML diagrams in the Software Factory! Blockdiag is ok-isch for simple diagrams but for anything slightly more complex, we should use PlantUML. (the rendering is also much more readable)
- A new Sequence diagram to distinguish between successful and fail updates
- A new appendix on how to generate artifacts from Yocto, download/compile/run Hawkbit locally and how to deploy a SWUpdate artifact to PELUX on a raspberrypi3.
- Fixes for most of your comments (only a couple of questions are left for Jonatan)

I think I'm done with everything I wanted to get merged.

P.S: Sorry for the large number of force push above, I needed some trial-and-error to get plantuml working on the CI system... :)

GordanM commented on Dockerfile:
```diff
@@ -5,12 +5,14 @@ RUN apt-get update && apt-get install -y \
 	python3-pip \
-	enchant
+	enchant \
+    plantuml
```
Nitpick: use same indent.

GordanM commented on docs/chapters/architecture/SOTA-system.rst:
```diff
@@ -0,0 +1,765 @@
+
+.. code-block:: bash
+
+    $ mvn clean install
```
Is there a dependencies chapter describing the dependencies of the server? Couldn't find one :)

FlorentRevest replied:
I added information about Java and Maven requirements in the "Context" subsection of this appendix. ;)

GordanM reviewed (approved):
Great job  @FlorentRevest ! It seems to me you've resolved quite a few topics with this PR.

Regarding your question with the commits to the dictionary, I would always recommend having a commit chain that is actually buildable and satisfies the CI. In case you need to revert some commits this will ensure we can pull parts out without breaking something. Although, you are right about the atomicity but that is something we should live with. :)
Also, I have nitpick comments.

thenor commented:
Regarding the commits, basically same as Gordan already wrote, I think the basic rule is that every commit should be buildable by itself, meaning that if words are introduced in one commit then they need to be included in the dict to be able to build.

GordanM commented:
@jonte do you have any further comments or can we merge this PR? :)

jonte commented:
I'm OK to merge.

GordanM commented:
@jonte thanks!

@tobsan and @thenor any last comments? @FlorentRevest I think you have a green light pretty much :).

FlorentRevest commented:
> @FlorentRevest I think you have a green light pretty much :).

Cool! I don't have the rights to merge it myself so I'll let you take care of that.
